### ISSUE TITLE
Nat deriving Ord problem

### ISSUE THREAD

**Issue by inexxt** [goldfirere/singletons#336]
When I try to compile the code
```haskell
$(singletons [d|
  data Nat = Zero | Succ Nat deriving (Show, Eq, Ord)
  pred' :: Nat -> Nat
  pred' Zero = Zero
  pred' (Succ n) = n
  |])
```
an error appears
```
/home/jack/opetopes-hs/src/Nattype.hs:32:3: error:
    Not in scope: type constructor or class ‘:$’
    Perhaps you meant one of these:
      ‘:<’ (imported from Data.Singletons.TH),
      ‘:>’ (imported from Data.Singletons.TH)
   |
32 | $(singletons [d|
```
while everything works correctly when remove `Ord` from the code above: 
 ```haskell
$(singletons [d|
  data Nat = Zero | Succ Nat deriving (Show, Eq)
  pred' :: Nat -> Nat
  pred' Zero = Zero
  pred' (Succ n) = n
  |])
```
 1) Why is that?
 2) If it is possible, it might be useful to change the error message - is not helpful at all, and impossible to google.

**Comment by mstksg** [comment]
What modules have you imported?  It is possible that you have not imported a module that exports `:$`.

As far as error messages go, this is actually one of the more common error messages you will encounter in Haskell:

    Not in scope: type constructor or class ????

This usually signifies that the identifier `????` is not in scope, because it has not been exported in most modules.  Usually the solution involves importing the module that defines `????`.

I'm not saying that the singletons library has perfect errors, but "Not in scope: type constructor or class ..." is one of the more ideal Haskell errors you can get, I feel :)  If you haven't run into it yet while using Haskell, it will actually pop up a lot in when using many different libraries as well, so it's somewhat ubiquitous.  (And I do see a lot of [advice using google](https://www.google.com/search?q=not+in+scope%3A+type+constructor+or+class) that suggests importing the module that the identifier is defined in)

**Comment by RyanGlScott** [comment]
I'm guessing you're using `singletons-2.3.*`? If so, that's due to an old issue with `Data.Singletons.TH` not reexporting everything that is used in derived instances. This has been fixed in `singletons-2.4.*` (see https://github.com/goldfirere/singletons/issues/190).

As @mstksg says, there's not much that can be done about these sorts of error messages in general in `singletons`, as the Template Haskell machinery must generate code where we can't know the module it lives in in advance. But in the particular case of derived code, we can at least reexport all of the functions that it might use from `Data.Singletons.TH`, which we have done in the most recent release.

I don't think there's much more that can be done here, so closing.
